ee/candi/cloud-providers/openstack/bashible/bundles/ubuntu-lts/bootstrap-networks.sh.tpl

```diff
@@ -5,38 +5,76 @@
-for mac in $configured_macs; do
-  ifname="$(echo "$ip_addr_show_output" | jq -re --arg mac "$mac" '.[] | select(.address == $mac) | .ifname')|"
-  configured_ifnames_pattern+="$ifname"
-done
-
-for i in /sys/class/net/!(${configured_ifnames_pattern%?}); do
-  if ! udevadm info "$i" 2>/dev/null | grep -Po '(?<=E: ID_NET_DRIVER=)virtio.*' 1>/dev/null 2>&1; then
-    continue
-  fi
+function is_ip_in_cidr() {
+  ip="$1"
+  IFS="/" read net_address net_prefix <<< "$2"
+
+  IFS=. read -r a b c d <<< "$ip"
+  ip_dec="$((a * 256 ** 3 + b * 256 ** 2 + c * 256 + d))"
+
+  IFS=. read -r a b c d <<< "$net_address"
+  net_address_dec="$((a * 256 ** 3 + b * 256 ** 2 + c * 256 + d))"
 
-  ifname=$(basename "$i")
-  mac="$(echo "$ip_addr_show_output" | jq -re --arg ifname "$ifname" '.[] | select(.ifname == $ifname) | .address')"
+  netmask=$(((0xFFFFFFFF << (32 - net_prefix)) & 0xFFFFFFFF))
 
-  cat > /etc/netplan/100-cim-"$ifname".yaml <<BOOTSTRAP_NETWORK_EOF
+  test $((netmask & ip_dec)) -eq $((netmask & net_address_dec))
+}
+
+function cat_file() {
+  dev=$1
+  metric=$2
+  mac=$3
+  cat > /etc/netplan/100-cim-"$dev".yaml <<BOOTSTRAP_NETWORK_EOF
 network:
   version: 2
   ethernets:
-    $ifname:
+    $cim_dev:
       dhcp4: true
       dhcp4-overrides:
         use-hostname: false
-        use-routes: false
+        route-metric: $metric
         use-dns: false
         use-ntp: false
       match:
         macaddress: $mac
 BOOTSTRAP_NETWORK_EOF
-done
+}
+
+ip_route_show_default_output=$(ip -json route show default)
+count_default=$(echo $ip_route_show_default_output | jq length)
+if [[ "$count_default" != "1" ]]; then
+  cim_dev=""
+  apiserverEndpoint={{ index (.normal.apiserverEndpoints | first | split ":") "_0" }}
+  ip_addr=$(ip -json addr)
+  for i in  { 1 .. $count_default }; do
+    i_dev=$(echo $ip_route_show_default_output | jq -r .[$i-1].dev)
+    i_addr_info=$(echo $ip_addr | jq '.[] | select(.ifname == "'$i_dev'") | .addr_info')
+    i_count_addr_info=$(echo $i_addr_info | jq length)
+    for j in { 1 .. $i_count_addr_info}; do
+      j_local=$(echo $i_addr_info | jq -r .[$j-1].local)
+      j_prefixlen=$(echo $i_addr_info | jq -r .[$j-1].prefixlen)
+      if is_ip_in_cidr "$apiserverEndpoint" "$j_local/$j_prefixlen"; then
+        cim_dev=$i_dev
+      fi
+    done
+  done
+  if [[ "$cim_dev" != "" ]]; then
+    metric=$(echo $ip_route_show_default_output | jq '.[] | select(.dev == "'$cim_dev'") | .metric')
+    for i in  { 1 .. $count_default }; do
+      i_dev=$(echo $ip_route_show_default_output | jq -r .[$i-1].dev)
+      if [[ "$i_dev" != "$cim_dev"]]; then
+        i_metric=$(echo $ip_route_show_default_output | jq '.[] | select(.dev == "'$i_dev'") | .metric')
+        if [[ "$i_metric" == "$metric" ]]; then
+          cim_mac=$(echo $ip_addr | jq -r '.[] | select(.ifname == "'$cim_dev'") | .address')
+          cim_metric=$(expr $metric + 100)
+          cat_file "$cim_dev" "$cim_metric" "$cim_mac"
```

Comment:
must be taken out of the loop to take into account the option with three default routes